File: SubstitutionCipher/substitution.c
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function prototypes
bool is_valid_key(string key);
string encrypt(string plaintext, string key);

int main(int argc, string argv[])
{
    // Check for single command-line argument
    if (argc != 2)
    {
        printf("Usage: ./substitution key\n");
        return 1;
    }

    string key = argv[1];

    // Validate the key
    if (!is_valid_key(key))
    {
        printf("Key must contain 26 unique alphabetic characters.\n");
        return 1;
    }

    // Prompt user for plaintext
    string plaintext = get_string("plaintext: ");

    // Encrypt the plaintext
    string ciphertext = encrypt(plaintext, key);

    // Output the ciphertext
    printf("ciphertext: %s\n", ciphertext);

    // Free the allocated memory for ciphertext
    free(ciphertext);

    return 0;
}

// Function to check if the key is valid
bool is_valid_key(string key)
{
    int length = strlen(key);

    // Key must be exactly 26 characters long
    if (length != 26)
    {
        return false;
    }

    bool seen[26] = {false};

    for (int i = 0; i < length; i++)
    {
        if (!isalpha(key[i]))
        {
            return false;
        }

        // Check for duplicate letters (case insensitive)
        int index = tolower(key[i]) - 'a';
        if (seen[index])
        {
            return false;
        }
        seen[index] = true;
    }

    return true;
}

// Function to encrypt plaintext using the substitution key
string encrypt(string plaintext, string key)
{
    int length = strlen(plaintext);
    string ciphertext = malloc(length + 1); // Allocate memory for ciphertext

    for (int i = 0; i < length; i++)
    {
        if (isalpha(plaintext[i]))
        {
            // Preserve case and substitute letters
            if (islower(plaintext[i]))
            {
                ciphertext[i] = tolower(key[plaintext[i] - 'a']);
            }
            else
            {
                ciphertext[i] = toupper(key[plaintext[i] - 'A']);
            }
        }
        else
        {
            // Non-alphabetic characters remain unchanged
            ciphertext[i] = plaintext[i];
        }
    }

    ciphertext[length] = '\0'; // Null
    return ciphertext;
}

File: SubstitutionCipher/substitution.cs
using System;

class Substitution
{
    // Main function
    static void Main(string[] args)
    {
        // Check for single command-line argument
        if (args.Length != 1)
        {
            Console.WriteLine("Usage: dotnet run <key>");
            Environment.Exit(1);
        }

        string key = args[0];

        // Validate the key
        if (!IsValidKey(key))
        {
            Console.WriteLine("Key must contain 26 unique alphabetic characters.");
            Environment.Exit(1);
        }

        // Prompt user for plaintext
        Console.Write("plaintext: ");
        string plaintext = Console.ReadLine();

        // Encrypt the plaintext
        string ciphertext = Encrypt(plaintext, key);

        // Output the ciphertext
        Console.WriteLine("ciphertext: " + ciphertext);
    }

    // Function to check if the key is valid
    static bool IsValidKey(string key)
    {
        if (key.Length != 26)
        {
            return false;
        }

        bool[] seen = new bool[26];

        for (int i = 0; i < key.Length; i++)
        {
            if (!char.IsLetter(key[i]))
            {
                return false;
            }

            int index = char.ToLower(key[i]) - 'a';
            if (seen[index])
            {
                return false;
            }
            seen[index] = true;
        }

        return true;
    }

    // Function to encrypt plaintext using the substitution key
    static string Encrypt(string plaintext, string key)
    {
        char[] ciphertext = new char[plaintext.Length];

        for (int i = 0; i < plaintext.Length; i++)
        {
            if (char.IsLetter(plaintext[i]))
            {
                if (char.IsLower(plaintext[i]))
                {
                    ciphertext[i] = char.ToLower(key[plaintext[i] - 'a']);
                }
                else
                {
                    ciphertext[i] = char.ToUpper(key[plaintext[i] - 'A']);
                }
            }
            else
            {
                // Non-alphabetic characters remain unchanged
                ciphertext[i] = plaintext[i];
            }
        }

        return new string(ciphertext);
    }
}
